core/drivers/crypto/stm32/stm32_hash.c

```diff
@@ -0,0 +1,909 @@
+	}
+
+	c->queue_size = next_queue_size;
+
+	if (len) {
+		assert(c->remain.len == 0);
+
+		if (!c->remain.buf) {
+			res = TEE_ERROR_BAD_STATE;
+			goto exit;
+		}
+
+		memcpy((uint8_t *)c->remain.buf, buffer, len);
+		c->remain.len = len;
+	}
+
+	res = save_context(c);
+
+exit:
+	clk_disable(c->dev->pdata.clock);
+	mutex_unlock(&c->dev->lock);
+
+	return res;
+}
+
+TEE_Result stm32_hash_final(struct stm32_hash_context *c, uint8_t *digest,
+			    const uint8_t *key, size_t len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	vaddr_t base = 0;
+
+	assert(c);
+
+	base = c->dev->pdata.base;
+
+	if ((!key || !len) && c->mode != STM32_HASH_MODE)
+		return TEE_ERROR_BAD_STATE;
+
+	mutex_lock(&c->dev->lock);
+	if (clk_enable(c->dev->pdata.clock)) {
+		EMSG("Fail to enable clk %s", c->dev->pdata.clock->name);
+		panic();
+	}
+
+	res = restore_context(c);
+	if (res)
+		goto exit;
+
+	if (c->remain.len) {
+		size_t i = 0;
+
+		if (!c->remain.buf) {
+			res = TEE_ERROR_BAD_STATE;
+			goto exit;
+		}
+
+		for (i = 0;
+		     i < ROUNDUP_DIV(c->remain.len, sizeof(uint32_t));
+		     i++) {
+			res = hash_write_data(base, c->remain.buf[i]);
+			if (res)
+				goto exit;
+			c->remain.buf[i] = 0; /* Reset to 0 */
+		}
+
+		io_clrsetbits32(base + _HASH_STR, _HASH_STR_NBLW_MASK,
+				8U * (c->remain.len % sizeof(uint32_t)));
+
+		/* No more saved data */
+		c->remain.len = 0;
+	} else {
+		io_clrbits32(base + _HASH_STR, _HASH_STR_NBLW_MASK);
+	}
+
+	io_setbits32(base + _HASH_STR, _HASH_STR_DCAL);
+
+	if (c->mode == STM32_HMAC_MODE) {
+		res = write_key(base, key, len);
+		if (res)
+			goto exit;
+	}
+
+	res = hash_get_digest(c, digest);
+
+exit:
+	clk_disable(c->dev->pdata.clock);
+	mutex_unlock(&c->dev->lock);
+
+	return res;
+}
+
+TEE_Result stm32_hash_init(struct stm32_hash_context *c, const uint8_t *key,
+			   size_t len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(c);
+
+	if ((!key || !len) && c->mode != STM32_HASH_MODE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&c->dev->lock);
+
+	if (clk_enable(c->dev->pdata.clock)) {
+		EMSG("Fail to enable clk %s", c->dev->pdata.clock->name);
+		panic();
+	}
+
+	c->remain.len = 0;
+	/* First queue is block_size + one register */
+	c->queue_size = c->block_size + sizeof(uint32_t);
+	memset(c->remain.buf, 0, c->queue_size);
+
+	res = hw_init(c, key, len);
+	if (res)
+		goto exit;
+
+	res = save_context(c);
+
+exit:
+	clk_disable(c->dev->pdata.clock);
+	mutex_unlock(&c->dev->lock);
+
+	return res;
+}
+
+static TEE_Result stm32_hash_parse_fdt(struct stm32_hash_platdata *pdata,
+				       const void *fdt, int node,
+				       const void *compat_data)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t reg_size = 0;
+	paddr_t reg = 0;
+
+	res = rstctrl_dt_get_by_index(fdt, node, 0, &pdata->reset);
+	if (res != TEE_SUCCESS && res != TEE_ERROR_ITEM_NOT_FOUND)
+		return res;
+
+	res = clk_dt_get_by_index(fdt, node, 0, &pdata->clock);
+	if (res)
+		return res;
+
+	res = fdt_reg_info(fdt, node, &reg, &reg_size);
```

Comment:
Maybe change base to be a `struct struct io_pa_va`

and use

```
if (fdt_reg_info(fdt, node, &pdata->base.pa, &reg_size))
	panic();
```

then

`io_pa_or_va_secure(&pdata->base, reg_size);`
?

I see that this is also the case in other crypto drivers so your choice